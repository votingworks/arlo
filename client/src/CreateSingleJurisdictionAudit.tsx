import React, { useState } from 'react'
import { toast } from 'react-toastify'
import { Formik, FormikProps, Field } from 'formik'
import { RouteComponentProps } from 'react-router-dom'
import styled from 'styled-components'
import FormSection from './components/Atoms/Form/FormSection'
import FormField from './components/Atoms/Form/FormField'
import { api } from './components/utilities'
import FormButton from './components/Atoms/Form/FormButton'

interface IValues {
  auditName: string
}

const Button = styled(FormButton)`
  margin: 30px 0 0 0;
`

const CenterField = styled(FormField)`
  width: 100%;
`

const Wrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  @media (min-width: 500px) {
    width: 400px;
  }
`

const CreateSingleJurisdictionAudit = ({ history }: RouteComponentProps) => {
  const [loading, setLoading] = useState(false)
  const onSubmit = async ({ auditName }: IValues) => {
    try {
      setLoading(true)
      const response: { electionId: string } = await api('/election/new', {
        method: 'POST',
        body: JSON.stringify({ auditName, isMultiJurisdiction: false }),
        headers: {
          'Content-Type': 'application/json',
        },
      })
      const { electionId } = response
      history.push(`/audit/${electionId}/setup`)
    } catch (err) /* istanbul ignore next */ {
      // will be removed when we migrate toasting to the api function
      toast.error(err.message)
      setLoading(false)
    }
  }
  return (
    <Wrapper>
      <img height="50px" src="/arlo.png" alt="Arlo, by VotingWorks" />
      <Formik onSubmit={onSubmit} initialValues={{ auditName: '' }}>
        {({ handleSubmit }: FormikProps<IValues>) => (
          <>
            <FormSection>
              {/* eslint-disable jsx-a11y/label-has-associated-control */}
              <label htmlFor="audit-name" id="audit-name-label">
                Give your new audit a unique name.
                <Field
                  id="audit-name"
                  aria-labelledby="audit-name-label"
                  name="auditName"
                  disabled={loading}
                  validate={(v: string) => (v ? undefined : 'Required')}
                  component={CenterField}
                />
              </label>
            </FormSection>
            <Button
              type="button"
              intent="primary"
              fill
              large
              onClick={handleSubmit}
              loading={loading}
              disabled={loading}
            >
              Create a New Audit
            </Button>
          </>
        )}
      </Formik>
    </Wrapper>
  )
}

export default CreateSingleJurisdictionAudit
