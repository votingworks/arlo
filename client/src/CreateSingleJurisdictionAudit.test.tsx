import React from 'react'
import { fireEvent, screen } from '@testing-library/react'
import { RouteComponentProps } from 'react-router-dom'
import CreateSingleJurisdictionAudit from './CreateSingleJurisdictionAudit'
import {
  routerTestProps,
  withMockFetch,
  renderWithRouter,
} from './components/testUtilities'

const routeProps: RouteComponentProps<{}> = routerTestProps('/audit', {})
routeProps.history.push = jest.fn()

const apiCalls = {
  putAudit: {
    url: '/api/election/new',
    options: {
      method: 'POST',
      body: JSON.stringify({
        auditName: 'Audit Name',
        isMultiJurisdiction: false,
      }),
      headers: {
        'Content-Type': 'application/json',
      },
    },
    response: { electionId: '12345' },
  },
}

const renderView = () =>
  renderWithRouter(<CreateSingleJurisdictionAudit {...routeProps} />, {
    route: '/audit',
  })

describe('CreateSingleJurisdictionAudit', () => {
  it('calls the /api/election/new endpoint for nonauthenticated user', async () => {
    const expectedCalls = [apiCalls.putAudit]
    await withMockFetch(expectedCalls, async () => {
      const { container } = renderView()

      fireEvent.change(
        await screen.findByLabelText('Give your new audit a unique name.'),
        {
          target: { value: 'Audit Name' },
        }
      )
      expect(container).toMatchSnapshot()

      fireEvent.click(screen.getByText('Create a New Audit'), { bubbles: true })

      await expect(routeProps.history.push).toHaveBeenCalledTimes(1)
    })
  })
})
