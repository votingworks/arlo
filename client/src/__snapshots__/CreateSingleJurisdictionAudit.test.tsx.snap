// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSingleJurisdictionAudit calls the /api/election/new endpoint for nonauthenticated user 1`] = `
<div>
  <div
    class="sc-gqjmRU goXkVq"
  >
    <img
      alt="Arlo, by VotingWorks"
      height="50px"
      src="/arlo.png"
    />
    <div
      class="sc-bdVaJa fdbLH"
    >
      <label
        for="audit-name"
        id="audit-name-label"
      >
        Give your new audit a unique name.
        <div
          class="sc-gZMcBi bddzxw sc-bxivhb dIUGFb"
        >
          <div
            class="bp3-input-group sc-ifAKCX juRnqx"
          >
            <input
              aria-labelledby="audit-name-label"
              class="bp3-input"
              id="audit-name"
              name="auditName"
              style="padding-right: 10px;"
              value="Audit Name"
            />
          </div>
        </div>
      </label>
    </div>
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-iwsKbI kljBkO sc-dnqmqq iQfLYL"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Create a New Audit
      </span>
    </button>
  </div>
</div>
`;
